### PR TITLE
feature add Meta Gateway for new design API

### PR DESCRIPTION
**Pull request by kpango**
### Description:
Implement meta gateway for new design API
this PR is sub PR for #742

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/gateway/meta/handler/grpc/handler.go [hunk 1]
```diff
@@ -0,0 +1,771 @@
...
+}
+
+func (s *server) Search(ctx context.Context, req *payload.Search_Request) (res *payload.Search_Response, err error) {
+	ctx, span := trace.StartSpan(ctx, apiName+".Search")
+	defer func() {
+		if span != nil {
+			span.End()
+		}
+	}()
+	if len(req.Vector) < 2 {
```
Comment (hlts2) [inline pkg/gateway/meta/handler/grpc/handler.go:88]:
```suggestion
	if len(req.GetVector) < 2 {
```

File: pkg/gateway/meta/usecase/vald.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+	var mech, sech, oech <-chan error
+	var err error
+	if r.observability != nil {
+		oech = r.observability.Start(ctx)
+	}
+	if r.metadata != nil {
+		mech, err = r.metadata.Start(ctx)
+		if err != nil {
+			close(ech)
+			return nil, err
```
Comment (hlts2) [inline pkg/gateway/meta/usecase/vald.go:172]: if `oech` is not nil, do you need the function to close `oech` too?

File: pkg/gateway/meta/handler/grpc/handler.go [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+		err = errors.ErrInvalidDimensionSize(len(req.Vector), 0)
```
Comment (hlts2) [inline pkg/gateway/meta/handler/grpc/handler.go:89]: same 
https://github.com/vdaas/vald/pull/746/files#r499426941

File: pkg/gateway/meta/router/router.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+				},
+				"/search/multi",
+				h.MultiSearch,
+			},
+			{
+				"Multi Search By ID",
+				[]string{
+					http.MethodGet,
+				},
+				"/search//multi/{id}",
```
Comment (rinx) [inline pkg/gateway/meta/router/router.go:82]: maybe?
```suggestion
				"/search/multi/{id}",
```

Comment (rinx) [reply pkg/gateway/meta/router/router.go:82]: or multi search by ID must be a `http.MethodPost`?

File: pkg/gateway/meta/service/meta.go [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+
+func (m *meta) Start(ctx context.Context) (<-chan error, error) {
+	if m.enableCache && m.cache != nil {
+		m.cache.Start(ctx)
+	}
+	return m.client.StartConnectionMonitor(ctx)
+}
+
+func (m *meta) Exists(ctx context.Context, meta string) (bool, error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.Exists")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:92]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.Exists")
```

File: pkg/gateway/meta/service/meta.go [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+
+	if m.enableCache {
+		m.cache.Set(uuidCacheKeyPref+meta, k)
+		m.cache.Set(metaCacheKeyPref+k, meta)
+	}
+	return true, nil
+}
+
+func (m *meta) GetMeta(ctx context.Context, uuid string) (v string, err error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.GetMeta")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:135]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.GetMeta")
```

File: pkg/gateway/meta/service/meta.go [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+
+	if m.enableCache {
+		m.cache.Set(metaCacheKeyPref+uuid, v)
+		m.cache.Set(uuidCacheKeyPref+v, uuid)
+	}
+	return v, nil
+}
+
+func (m *meta) GetMetas(ctx context.Context, uuids ...string) ([]string, error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.GetMetas")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:171]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.GetMetas")
```

File: pkg/gateway/meta/service/meta.go [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+				}
+			}
+			return vs, err
+		}
+	}
+	return nil, err
+}
+
+func (m *meta) GetUUID(ctx context.Context, meta string) (k string, err error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.GetUUID")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:220]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.GetUUID")
```

File: pkg/gateway/meta/service/meta.go [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+	k = key.(string)
+	if m.enableCache {
+		m.cache.Set(uuidCacheKeyPref+meta, k)
+		m.cache.Set(metaCacheKeyPref+k, meta)
+	}
+	return k, nil
+}
+
+func (m *meta) GetUUIDs(ctx context.Context, metas ...string) ([]string, error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.GetUUIDs")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:256]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.GetUUIDs")
```

File: pkg/gateway/meta/service/meta.go [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+				}
+			}
+			return ks, err
+		}
+	}
+	return nil, err
+}
+
+func (m *meta) SetUUIDandMeta(ctx context.Context, uuid, meta string) (err error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.SetUUIDandMeta")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:305]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.SetUUIDandMeta")
```

File: pkg/gateway/meta/service/meta.go [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+
+	if m.enableCache {
+		m.cache.Set(uuidCacheKeyPref+meta, uuid)
+		m.cache.Set(metaCacheKeyPref+uuid, meta)
+	}
+	return nil
+}
+
+func (m *meta) SetUUIDandMetas(ctx context.Context, kvs map[string]string) (err error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.SetUUIDandMetas")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:333]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.SetUUIDandMetas")
```

File: pkg/gateway/meta/service/meta.go [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+		for uuid, meta := range kvs {
+			m.cache.Set(uuidCacheKeyPref+meta, uuid)
+			m.cache.Set(metaCacheKeyPref+uuid, meta)
+		}
+	}
+	return nil
+}
+
+func (m *meta) DeleteMeta(ctx context.Context, uuid string) (v string, err error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.DeleteMeta")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:369]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.DeleteMeta")
```

File: pkg/gateway/meta/service/meta.go [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+		return val.GetVal(), nil
+	})
+	if err != nil {
+		return "", err
+	}
+	return val.(string), nil
+}
+
+func (m *meta) DeleteMetas(ctx context.Context, uuids ...string) ([]string, error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.DeleteMetas")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:399]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.DeleteMetas")
```

File: pkg/gateway/meta/service/meta.go [hunk 14]
```diff
@@ -0,0 +1,490 @@
...
+		return vals.GetVals(), nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return vals.([]string), nil
+}
+
+func (m *meta) DeleteUUID(ctx context.Context, meta string) (string, error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.DeleteUUID")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:431]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.DeleteUUID")
```

File: pkg/gateway/meta/service/meta.go [hunk 15]
```diff
@@ -0,0 +1,490 @@
...
+		return key.GetKey(), nil
+	})
+	if err != nil {
+		return "", err
+	}
+	return key.(string), nil
+}
+
+func (m *meta) DeleteUUIDs(ctx context.Context, metas ...string) ([]string, error) {
+	ctx, span := trace.StartSpan(ctx, "vald/gateway-vald/service/Meta.DeleteUUIDs")
```
Comment (rinx) [inline pkg/gateway/meta/service/meta.go:461]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "vald/gateway-meta/service/Meta.DeleteUUIDs")
```

File: pkg/gateway/meta/handler/grpc/option.go [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+type Option func(*server)
+
+var (
+	defaultOpts = []Option{
+		WithErrGroup(errgroup.Get()),
+		WithStreamConcurrency(20),
+	}
+)
+
+func WithClient(g vald.Client) Option {
```
Comment (rinx) [inline pkg/gateway/meta/handler/grpc/option.go:35]: WithClient is a too abstract name i think. how about to name it as WithValdClient?

File: .github/workflows/dockers-gateway-meta-image.yml [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+      with:
+        registry: ghcr.io
+        username: ${{ secrets.PACKAGE_USER }}
+        password: ${{ secrets.PACKAGE_TOKEN }}
+    - name: Image name
+      id: image_name
+      run: |
+        image_name=`make docker/name/gateway-meta`
+        base_platforms=`make docker/platforms`
+        echo "::set-env name=IMAGE_NAME::${image_name}"
```
Comment (rinx) [inline .github/workflows/dockers-gateway-meta-image.yml:85]:
```suggestion
        echo "IMAGE_NAME=${image_name}" >> $GITHUB_ENV
```

File: .github/workflows/dockers-gateway-meta-image.yml [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+        echo "::set-output name=IMAGE_NAME::${image_name}"
+        echo "::set-output name=BASE_PLATFORMS::${base_platforms}"
+    - name: Determine tag name (master)
+      if: github.ref == 'refs/heads/master'
+      run: |
+        echo "::set-env name=PRIMARY_TAG::nightly"
+        echo "::set-env name=PLATFORMS::${PLATFORMS}"
```
Comment (rinx) [inline .github/workflows/dockers-gateway-meta-image.yml:92]:
```suggestion
        echo "PRIMARY_TAG=nightly" >> $GITHUB_ENV
        echo "PLATFORMS=${PLATFORMS}" >> $GITHUB_ENV
```

File: .github/workflows/dockers-gateway-meta-image.yml [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+      env:
+        PLATFORMS: linux/amd64,linux/arm64
+    - name: Determine tag name (pull request)
+      if: github.event_name == 'pull_request'
+      run: |
+        pr_num=`cat $GITHUB_EVENT_PATH | jq -r ".number"`
+        echo "PR-${pr_num}" > versions/VALD_VERSION
+        echo "::set-env name=PRIMARY_TAG::pr-${pr_num}"
+        echo "::set-env name=PLATFORMS::${PLATFORMS}"
```
Comment (rinx) [inline .github/workflows/dockers-gateway-meta-image.yml:101]:
```suggestion
        echo "PRIMARY_TAG=pr-${pr_num}" >> $GITHUB_ENV
        echo "PLATFORMS=${PLATFORMS}" >> $GITHUB_ENV
```

File: .github/workflows/dockers-gateway-meta-image.yml [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+      env:
+        PLATFORMS: linux/amd64
+    - name: Determine tag name (tags)
+      if: startsWith( github.ref, 'refs/tags/')
+      id: determine_tag
+      run: |
+        tag_name=`echo $GITHUB_REF | sed -e 's:^refs/tags/::'`
+        echo "::set-output name=TAG_NAME::${tag_name}"
+        echo "::set-env name=PRIMARY_TAG::${tag_name}"
+        echo "::set-env name=PLATFORMS::${PLATFORMS}"
```
Comment (rinx) [inline .github/workflows/dockers-gateway-meta-image.yml:111]:
```suggestion
        echo "PRIMARY_TAG=${tag_name}" >> $GITHUB_ENV
        echo "PLATFORMS=${PLATFORMS}" >> $GITHUB_ENV
```

**Review by rinx — COMMENTED** [review]
These changes are releated to #752.

File: pkg/gateway/meta/usecase/vald.go (hunk 2 above)
Comment (kpango) [reply pkg/gateway/meta/usecase/vald.go:172]: oech is read only channel we can't handle here.

**Comment by kpango** [comment]
@rinx @hlts2 resolved.

**Comment by rinx** [comment]
ok

**Review by hlts2 — APPROVED** [review]
LGTM

**Pull request merged**